ext/tiny_tds/client.c

```diff
@@ -57,6 +57,15 @@ VALUE rb_tinytds_raise_error(DBPROCESS *dbproc, int is_message, int cancel, cons
 
 
 // Lib Backend (Memory Management & Handlers)
+static void push_userdata_error(tinytds_client_userdata *userdata, tinytds_errordata error) {
+  // reallocate memory for the array as needed
+  if (userdata->nonblocking_errors_size == userdata->nonblocking_errors_length) {
+    userdata->nonblocking_errors_size *= 2;
+    userdata->nonblocking_errors = realloc(userdata->nonblocking_errors, userdata->nonblocking_errors_size * sizeof(tinytds_errordata));
+  }
```

Comment:
This seems odd, so I have a few questions (not a pro with C, these are real questions)

1. When would `push_userdata_error` be called, and this if statement _not_ be true. I'd expect that every time we need to push an error on to the array, we'd need to resize the array.
2. Why is `nonblocking_errors_size` being doubled every time? Wouldn't it be preferable to incrementally grow the size in multiples of `sizeof(tinytds_errordata)`?
3. Why isn't the realloc like count x size, eg `realloc(userdata->nonblocking_errors, (userdata->nonblocking_errors_length + 1) * sizeof(tinytds_errordata))`?
3. Instead of tracking more state as a struct field, is anything gained by doing `sizeof(userdata->nonblocking_errors)` (assuming this would give the right answer, and if not why not?)

Reply:
If we attempt to push an error to the array and we don't have any more allocated space to add the new error we double the allocated size for the array. This allocates space for the new error as well as _future_ space that we may end up using. This is pretty standard practice for creating dynamic arrays. The reason being that reallocating the new space is expensive (`O(n)`) since the memory may have to be copied somewhere else in order to find the contiguous space for the new array size. As a result we want to reduce the number of times we have to do that operation.

1. `nonblocking_errors_size` represents the current _allocated_ size of the array. So these values are only ever the same if we've used up all of the currently allocated space.
2. The cost of memory allocation. See above.
3. The cost of memory allocation. See above.
4. `sizeof(userdata->nonblocking_errors)` would report the size of the pointer itself _not_ the size of the array that the pointer points to. So it wouldn't give the answer you're expecting.

Reply:
Alright, thanks for explaining. I guess it makes sense to optimize for minimal number of allocation _operations_ rather than _total memory size_, as a strategy. I will defer to your judgement on that this is both the right strategy and the appropriate implementation of that strategy.